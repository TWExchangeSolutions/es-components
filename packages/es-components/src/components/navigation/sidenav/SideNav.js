import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import uncontrollable from 'uncontrollable';
import viaTheme from 'es-components-via-theme';

import NavItem from './NavItem';

const NavStyled = styled.nav`
  background-color: ${props => props.theme.colors.white};
  box-shadow: 2px 2px 5px ${props => props.theme.colors.grayDark};

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }
`;

export const SideNav = props => {
  const { useAltStyle, children, onItemSelected, selected, theme } = props;

  const onNavClick = (navId = null) => {
    onItemSelected(navId);
  };

  return (
    <ThemeProvider theme={theme}>
      <NavStyled>
        <ul>
          {Children.toArray(children).map(child => {
            if (child !== null && child.type === NavItem) {
              const currentSelected = selected;
              return cloneElement(child, {
                useAltStyle,
                highlightedId: currentSelected,
                onNavClick
              });
            }
            return child;
          })}
        </ul>
      </NavStyled>
    </ThemeProvider>
  );
};

SideNav.propTypes = {
  /** Use the alternate nav style */
  useAltStyle: PropTypes.bool,
  children: PropTypes.node,
  /** Set the selected nav item by id, controlled mode */
  selected: PropTypes.string,
  /** Use to set a default nav, uncontrolled mode */
  defaultSelected: PropTypes.string,
  /** Function called when a nav item is clicked */
  onItemSelected: PropTypes.func,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

SideNav.defaultProps = {
  useAltStyle: false,
  onItemSelected: noop,
  theme: viaTheme
};

const UncontrolledSideNav = uncontrollable(SideNav, {
  selected: 'onItemSelected'
});

UncontrolledSideNav.Item = NavItem;

export default withTheme(UncontrolledSideNav);
