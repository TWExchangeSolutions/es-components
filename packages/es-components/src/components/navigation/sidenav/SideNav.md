(Uncontrolled) Use `defaultSelected` to set an initial selected item.
```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home" targetUrl="#home">Home</SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info">About</SideNav.Item>
      <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav useAltStyle defaultSelected="home">
      <SideNav.Item id="home">Home</SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info">About</SideNav.Item>
      <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>
</div>
```

(Controlled) Use `onItemSelected` to control the `selected` value.
```
class NavExample extends React.Component {
  constructor() {
    this.state = { selected: "home" };
    this.onItemSelected = this.onItemSelected.bind(this);
  }

  onItemSelected(value) {
    this.setState({ selected: value });
    alert(`${value} selected!`);
  };

  render() {
    return (
      <div style={{width: '30%'}}>
        <SideNav selected={ this.state.selected } onItemSelected={ this.onItemSelected }>
          <SideNav.Item id="home"><Icon name="home" /> Home</SideNav.Item>
          <SideNav.Item id="cart"><Icon name="shopping-cart" /> Cart</SideNav.Item>
          <SideNav.Item id="disabled" isDisabled><Icon name="ban-circle" /> Disabled</SideNav.Item>
        </SideNav>
      </div>
    )
  }
}
<NavExample />
```

Set `onClick` on an individual `SideNav.Item`.
```
<div style={{width: '30%'}}>
  <SideNav useAltStyle>
    <SideNav.Item id="home" onClick={ () => (alert('Time to go Home!'))}>Home</SideNav.Item>
    <SideNav.Item id="cart" onClick={ () => (alert('You clicked the Cart!'))}>Cart</SideNav.Item>
    <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Assign a `targetUrl` to set the nav link href location. Use `isExternalLink` to open the link in a new browser window.
```
<div style={{width: '30%'}}>
  <SideNav>
    <SideNav.Item id="home" targetUrl="/">Home</SideNav.Item>
    <SideNav.Item id="medicare" targetUrl="https://medicare.oneexchange.com" isExternalLink>Medicare Site</SideNav.Item>
    <SideNav.Item id="sidenav" targetUrl="#sidenav">SideNav Section</SideNav.Item>
  </SideNav>
</div>
```
