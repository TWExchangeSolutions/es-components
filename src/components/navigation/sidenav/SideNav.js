import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../theme';
import styled from 'styled-components';
import { noop } from 'lodash';

import NavItem from './NavItem';

const NavStyled = styled.nav`
  background-color: ${colors.white};
  box-shadow: 2px 2px 5px ${colors.grayDark};

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }
`;

const SideNav = props => {
  const { altStyle, children, onItemSelected, selected } = props;

  const onNavClick = (navId = null) => {
    onItemSelected(navId);
  };

  return (
    <NavStyled>
      <ul>
        {Children.toArray(children).map(child => {
          if (child !== null && child.type === NavItem) {
            const currentSelected = selected;
            return cloneElement(child, {
              altStyle,
              highlightedId: currentSelected,
              onNavClick
            });
          }
          return child;
        })}
      </ul>
    </NavStyled>
  );
};

SideNav.propTypes = {
  /** Use the alternate nav style */
  altStyle: PropTypes.bool,
  children: PropTypes.node,
  /** Use to manually select nav item by id, controlled mode */
  selected: PropTypes.string,
  /** Use to set a default nav, uncontrolled mode */
  defaultSelected: PropTypes.string,
  /** Function called when a nav item is clicked */
  onItemSelected: PropTypes.func
};

SideNav.defaultProps = {
  altStyle: false,
  onItemSelected: noop
};

SideNav.Item = NavItem;

export default SideNav;
