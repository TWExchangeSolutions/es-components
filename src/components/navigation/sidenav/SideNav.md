```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home" targetUrl="#home">Home</SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info">About</SideNav.Item>
      <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav defaultSelected="home" altStyle>
      <SideNav.Item id="home">Home</SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info">About</SideNav.Item>
      <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>
</div>
```

Assign a `targetUrl` to set the nav link href location. Use `isExternalLink` to open the link in a new browser window.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home">
    <SideNav.Item id="home" targetUrl="/" onNavClick={ () => (alert('Time to go Home!'))}>Home</SideNav.Item>
    <SideNav.Item id="medicare" targetUrl="https://medicare.oneexchange.com" isExternalLink>Medicare Site</SideNav.Item>
    <SideNav.Item id="alerts" targetUrl="#alert">Alert Section</SideNav.Item>
  </SideNav>
</div>
```

Setting `onItemSelected` will execute the function when a `SideNav.Item` is clicked.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" onItemSelected={ (id) => {alert(id); }}>
    <SideNav.Item id="home"><Icon name="home" /> Home</SideNav.Item>
    <SideNav.Item id="cart"><Icon name="shopping-cart" /> Cart</SideNav.Item>
    <SideNav.Item id="disabled" isDisabled><Icon name="ban-circle" /> Disabled</SideNav.Item>
  </SideNav>
</div>
```

Set `onNavClick` on an individual `SideNav.Item`.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" altStyle>
    <SideNav.Item id="home" onNavClick={ () => (alert('Time to go Home!'))}>Home</SideNav.Item>
    <SideNav.Item id="cart" onNavClick={ () => (alert('You clicked the Cart!'))}>Cart</SideNav.Item>
    <SideNav.Item id="disabled" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Manage selected item state manually using `selected` instead of `defaultSelected`.
```
<div style={{width: '30%'}}>
  <SideNav selected={state.selected} onItemSelected={ (id) => {setState({selected: id})} }>
    <SideNav.Item id="home">Home</SideNav.Item>
    <SideNav.Item id="cart">Cart</SideNav.Item>
    <SideNav.Item id="help">Help</SideNav.Item>
  </SideNav>
</div>
```
