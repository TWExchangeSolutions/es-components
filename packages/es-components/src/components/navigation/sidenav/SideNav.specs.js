/* eslint-env jest */

import React from 'react';
import { mountWithTheme, renderWithTheme } from '../../../testing';
import SideNav from './SideNav';

describe('drawer', () => {
  let instanceToRender;
  const onItemSelected = jest.fn();
  const onClick = jest.fn();

  beforeEach(() => {
    onItemSelected.mockClear();
    onClick.mockClear();

    instanceToRender = (
      <SideNav
        onItemSelected={navId => {
          onItemSelected(navId);
        }}
      >
        <SideNav.Item
          id="home"
          className="home"
          onClick={navId => onClick(navId)}
          targetUrl="/home"
        >
          Home
        </SideNav.Item>
        <SideNav.Item id="cart" className="cart">
          Cart
        </SideNav.Item>
        <SideNav.Item
          id="external"
          className="external"
          targetUrl="http://www.google.com"
          isExternalLink
        >
          External Link
        </SideNav.Item>
        <SideNav.Item
          id="info"
          className="info"
          onClick={() => onClick()}
          isDisabled
        >
          Disabled
        </SideNav.Item>
      </SideNav>
    );
  });

  it('renders as expected', () => {
    const tree = renderWithTheme(instanceToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('executes onItemSelected with the id of the nav item clicked', () => {
    const navInstance = mountWithTheme(instanceToRender);
    const navItem = navInstance.find('.cart').hostNodes();

    navItem.simulate('click');
    expect(onItemSelected).toBeCalledWith('cart');
  });

  it('executes onClick when nav item clicked', () => {
    const navInstance = mountWithTheme(instanceToRender);
    const navItem = navInstance.find('.home').hostNodes();

    navItem.simulate('click');
    expect(onClick).toBeCalledWith('home');
  });

  it('disabled item prevents onclick functions', () => {
    const navInstance = mountWithTheme(instanceToRender);
    const navItem = navInstance.find('.info').hostNodes();

    navItem.simulate('click');
    expect(onClick).not.toBeCalled();
    expect(onItemSelected).not.toBeCalled();
  });
});
