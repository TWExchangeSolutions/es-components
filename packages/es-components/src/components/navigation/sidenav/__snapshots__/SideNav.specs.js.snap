// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <nav
    class="es-sidenav sc-ifAKCX eEDsWM"
  >
    <ul>
      <li
        class="es-sidenav__navitem"
      >
        <a
          class="home sc-bwzfXH guxetE"
          href="/home"
        >
          <span
            class="sc-bxivhb dgmReD"
          >
            Home
          </span>
          <i
            aria-hidden="true"
            class="es-icon sc-bdVaJa tXzOW"
            content="61f"
            font-family="indv-mkt-icons"
            font-size="inherit"
          />
        </a>
      </li>
      <li
        class="es-sidenav__navitem"
      >
        <a
          class="cart sc-bwzfXH guxetE"
          href="#/"
        >
          <span
            class="sc-bxivhb dgmReD"
          >
            Cart
          </span>
          <i
            aria-hidden="true"
            class="es-icon sc-bdVaJa tXzOW"
            content="61f"
            font-family="indv-mkt-icons"
            font-size="inherit"
          />
        </a>
      </li>
      <li
        class="es-sidenav__navitem"
      >
        <a
          class="external sc-bwzfXH guxetE"
          href="http://www.google.com"
          target="_blank"
        >
          <span
            class="sc-bxivhb dgmReD"
          >
            External Link
          </span>
          <i
            aria-hidden="true"
            class="es-icon sc-bdVaJa tXzOW"
            content="61f"
            font-family="indv-mkt-icons"
            font-size="inherit"
          />
        </a>
      </li>
      <li
        class="es-sidenav__navitem"
      >
        <a
          class="info sc-bwzfXH kPjAmK"
          href="#/"
        >
          <span
            class="sc-bxivhb dgmReD"
          >
            Disabled
          </span>
          <i
            aria-hidden="true"
            class="es-icon sc-bdVaJa tXzOW"
            content="61f"
            font-family="indv-mkt-icons"
            font-size="inherit"
          />
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
